pkg/relay/register.go

```diff
@@ -0,0 +1,296 @@
+package relay
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sync/atomic"
+
+	"github.com/blocknative/dreamboat/pkg/structs"
+	"github.com/flashbots/go-boost-utils/types"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	ResponseTypeVerify = iota
+	ResponseTypeOthers
+	ResponseTypeStored
+)
+
+type TimestampRegistry interface {
+	Get(pubkey string) (timestamp uint64, ok bool)
+}
+
+// VerifyReq is a request structure used in communication
+// between api calls and fixed set of worker goroutines
+// it's using return channel pattern, meaning that after
+// sent the sender locks on that channel to get the response
+type VerifyReq struct {
+	Signature [96]byte
+	Pubkey    [48]byte
+	Msg       [32]byte
+	// Unique identifier of payload
+	// if needed to be passed back in response
+	ID       int
+	Response chan Resp
+}
+
+// StoreReq is similar to VerifyReq jsut for storing payloads
+type StoreReq struct {
+	RawPayload json.RawMessage
+	Pubkey     types.PublicKey
+	ID         int
+	Response   chan Resp
+}
+
+// Resp respone structure
+// - potential candidate for structure pool
+// as it's almost constant size
+type Resp struct {
+	ID     int
+	Type   int8
+	Commit bool
+	Err    error
+}
+
+// ***** Builder Domain *****
+// RegisterValidator is called is called by validators communicating through mev-boost who would like to receive a block from us when their slot is scheduled
+func (rs *Relay) RegisterValidator(ctx context.Context, payload []structs.SignedValidatorRegistration) error {
+	timer := prometheus.NewTimer(rs.m.Timing.WithLabelValues("registerValidator", "all"))
+	defer timer.ObserveDuration()
+
+	var respCh chan Resp
+
+	// This function is limitted to the size of the buffer to prevent deadlocks
+	if rs.config.RegisterValidatorMaxNum > uint64(len(payload)) {
+		respCh = rs.retChannPool.Get().(chan Resp)
+		defer rs.retChannPool.Put(respCh)
+	}
+
+	fc := NewFlowControl(respCh, len(payload))
```

Comment:
Can we add a comment clarifying that it's OK for `respCh` to be `nil`?  I had to double-check the implementation of `NewFlowControl` to be certain.  Additionally, can we add a docstring to `NewFlowControl` that declares a contract for how nil channels are handled?

Reply:
Added comment

Reply:
Nevermind, I see it now.  GitHub is so weird sometimes...